File: src/main/scala/io/octopus/model/Status.scala
package io.octopus.model

object Status {
  val IDLE    = "IDLE" 
  val RUNNING = "RUNNING" 
  val SUCCESS = "SUCCESS"
  val ERROR   = "ERROR"
  val KILLED  = "KILLED"
  val PAUSED  = "PAUSED"
  val BLOCKED = "BLOCKED"  
  val DONE    = "DONE"

  val valid = Map( 
    IDLE -> List(IDLE,RUNNING,DONE,BLOCKED),
    RUNNING -> List(ERROR, KILLED,SUCCESS),
    SUCCESS -> List(IDLE,RUNNING),
    ERROR -> List(IDLE,RUNNING,DONE),
    KILLED -> List(KILLED,IDLE,RUNNING),
    PAUSED -> List(PAUSED,IDLE,RUNNING),
    BLOCKED -> List(BLOCKED,IDLE),
    DONE -> List(DONE,IDLE,ERROR,BLOCKED,KILLED)
  )  

  def isValid(from: String, to: String): Boolean = valid.get(from).get.contains(to)

}

File: src/main/resources/static/js/constants/Status.js
define([
],function(){

  const IDLE  = 'IDLE',
      DONE    = 'DONE',
      SUCCESS = 'SUCCESS', 
      ERROR   = 'ERROR', 
      KILLED  = 'KILLED',
      BLOCKED = 'BLOCKED',
      RUNNING = 'RUNNING';

  const progress = (instance) => {

    if([SUCCESS, DONE].indexOf(instance.status) != -1 ){
      return 100;
    }
    if([ERROR, IDLE, KILLED, BLOCKED].indexOf(instance.status) != -1 ){
      return 0;
    }
    
    if(instance.task.stats == null || instance.task.stats.avgDuration == null || instance.task.stats.avgDuration == 0){
      Math.floor( Math.random() * 100 );
    }

    if(instance.task.stats == null){
      return 0;
    }

    return Math.floor(100*moment
      .duration(moment(new Date())
      .diff(moment(instance.startDate)))
      .asSeconds()/instance.task.stats.avgDuration);

  };   

  return {
    IDLE, 
    DONE,   
    SUCCESS,
    ERROR,  
    KILLED, 
    BLOCKED,
    RUNNING,
    progress
  }
});